File: Books/A_Comprehensive_Approach_to_ML_and_AI/python/manifold_learning_tutorial.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.spatial.distance import pdist, squareform
from scipy.linalg import eigh

def myLLE(X, K, d):
    # X: (N x D) data matrix, K: number of neighbors, d: target dimension.
    N = X.shape[0]
    dist = squareform(pdist(X))
    
    W = np.zeros((N, N))
    for i in range(N):
        # Get indices of K nearest neighbors (excluding self)
        idx = np.argsort(dist[i, :])
        neighbors = idx[1:K+1]
        # Compute local covariance matrix
        Z = (X[i, :] - X[neighbors, :]).T  # shape (D, K)
        C = Z.T @ Z
        # Regularization for numerical stability
        C += np.eye(K) * 1e-3 * np.trace(C)
        # Solve for weights
        w = np.linalg.solve(C, np.ones((K, 1)))
        w = w / np.sum(w)
        W[i, neighbors] = w.flatten()
    
    M = (np.eye(N) - W).T @ (np.eye(N) - W)
    # Compute eigen-decomposition (use eigh for symmetric matrices)
    eigvals, eigvecs = eigh(M)
    # Sort eigenvectors by ascending eigenvalues
    idx = np.argsort(eigvals)
    eigvecs = eigvecs[:, idx]
    # Discard the first eigenvector (zero eigenvalue)
    Y = eigvecs[:, 1:d+1]
    
    if d >= 2:
        plt.figure()
        plt.scatter(Y[:, 0], Y[:, 1], c='b', s=25)
        plt.title("LLE Embedding (First Two Dimensions)")
        plt.xlabel("Dimension 1")
        plt.ylabel("Dimension 2")
        plt.grid(True)
        plt.show()
    return Y

def generate_swiss_roll(N=1000):
    # Generate a Swiss Roll dataset.
    t = (3*np.pi/2) * (1 + 2 * np.random.rand(N))
    h = 21 * np.random.rand(N)
    X = np.zeros((N, 3))
    X[:, 0] = t * np.cos(t)
    X[:, 1] = h
    X[:, 2] = t * np.sin(t)
    
    fig = plt.figure()
    ax = fig.add_subplot(111, projection='3d')
    sc = ax.scatter(X[:, 0], X[:, 1], X[:, 2], c=t, cmap='viridis', s=20)
    ax.view_init(elev=6, azim=-17)
    plt.title("Swiss Roll Dataset")
    ax.set_xlabel("X")
    ax.set_ylabel("Y")
    ax.set_zlabel("Z")
    plt.colorbar(sc)
    plt.show()
    
    return X, t

# Main script for manifold learning
if __name__ == '__main__':
    X, t_param = generate_swiss_roll(1000)
    # Run LLE with K=12 neighbors and target dimension 2
    Y_embed = myLLE(X, K=12, d=2)

File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/manifold_learning_tutorial.m
close all; clear all; clc; 

% Load or generate a dataset (e.g., a Swiss Roll)
[X, t] = generateSwissRoll(1000); % Assume a function that generates Swiss Roll data
Y = myLLE(X, 12, 2);

function Y = myLLE(X, K, d)
    % X: Data matrix of size N x D
    % K: Number of nearest neighbors
    % d: Target dimensionality
    N = size(X,1);
    
    % Compute pairwise distances
    dist = squareform(pdist(X));
    
    % Initialize weight matrix W (N x N)
    W = zeros(N, N);
    
    for i = 1:N
        % Find the indices of K nearest neighbors (excluding self)
        [~, idx] = sort(dist(i,:));
        neighbors = idx(2:K+1);
        
        % Compute the local covariance matrix
        Z = X(i,:) - X(neighbors,:);
        C = Z * Z';
        
        % Regularization (for numerical stability)
        C = C + eye(K)*1e-3*trace(C);
        
        % Solve for weights that minimize the reconstruction error
        w = C \ ones(K, 1);
        w = w / sum(w);
        
        % Assign the weights to the global weight matrix
        W(i, neighbors) = w';
    end
    
    % Compute the matrix M = (I - W)'*(I - W)
    M = (eye(N) - W)' * (eye(N) - W);
    
    % Compute eigenvalues and eigenvectors
    [eigvec, eigval] = eig(M);
    
    % Sort eigenvalues in ascending order
    [~, idx] = sort(diag(eigval));
    eigvec = eigvec(:, idx);
    
    % Discard the first eigenvector (corresponding to zero eigenvalue)
    Y = eigvec(:, 2:d+1);
    
    % Visualization: Plot first two dimensions if d >= 2
    if d >= 2
        figure;
        scatter(Y(:,1), Y(:,2), 25, 'filled');
        title('LLE Embedding (First Two Dimensions)');
        xlabel('Dimension 1'); ylabel('Dimension 2');
        grid on;
    end

    % save_all_figs_OPTION('results/manifoldLearning2','png',1)
end

function [X, t] = generateSwissRoll(N)
% generateSwissRoll Generate a Swiss Roll dataset.
%   [X, t] = generateSwissRoll(N) generates a Swiss Roll with N data points.
%   X is an N-by-3 matrix representing 3D coordinates of the Swiss Roll,
%   and t is the parameter vector that can be used to color or order the points.
%
%   The Swiss Roll is constructed by sampling t uniformly from an interval
%   and computing the corresponding (x, y, z) coordinates as follows:
%       x = t * cos(t)
%       y = h    (random height component)
%       z = t * sin(t)
%
%   Example:
%       [X, t] = generateSwissRoll(1000);
%       scatter3(X(:,1), X(:,2), X(:,3), 20, t, 'filled');
%       title('Swiss Roll Dataset');
%       xlabel('X'); ylabel('Y'); zlabel('Z');
%       colorbar;
%
%   Inputs:
%       N - Number of data points (default: 1000)
%
%   Outputs:
%       X - An N-by-3 matrix containing the Swiss Roll coordinates
%       t - A vector of length N with the underlying parameter values

    if nargin < 1
        N = 1000;  % Default number of points
    end

    % Generate the parameter t uniformly in the interval [1.5*pi, 4.5*pi]
    t = (3*pi/2) * (1 + 2 * rand(N, 1));

    % Generate a height component uniformly in [0, 21]
    h = 21 * rand(N, 1);

    % Compute the 3D coordinates of the Swiss Roll
    X(:,1) = t .* cos(t);  % X-coordinate
    X(:,2) = h;            % Y-coordinate (height)
    X(:,3) = t .* sin(t);  % Z-coordinate

    figure;
    scatter3(X(:,1), X(:,2), X(:,3), 20, t, 'filled');
    view(-17,6) %az=-17,el=6
    title('Swiss Roll Dataset');
    xlabel('X'); ylabel('Y'); zlabel('Z');
    colorbar;
    
    % save_all_figs_OPTION('results/manifoldLearning1','png',1)
end

